sopel/modules/safety.py

```diff
@@ -122,19 +113,67 @@ def update_local_cache(bot: Sopel, init: bool = False):
             # Python on Windows throws an exception if the file is in use
             LOGGER.info('Could not delete %s: %s', old_file, str(err))
 
-    loc = os.path.join(bot.settings.homedir, "unsafedomains.txt")
-    if not os.path.isfile(loc) or os.path.getmtime(loc) < time.time() - 24 * 60 * 60:
-        # File doesn't exist or is older than one day — update it
-        url = bot.settings.safety.domain_blocklist_url
-        if url is None or not url.startswith("http"):
-            url = "https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts"
-        LOGGER.info("Downloading malicious domain list from %s", url)
-        # TODO: Can we use a cache header to avoid the download if it's unmodified?
-        urlretrieve(url, loc)
-    elif not init:
+    update_local_cache(bot, init=True)
+
+
+def zwsp(text: str) -> str:
+    """Insert a zero-width space between each character.
+
+    Useful for reducing highlights and clickability of links.
+    """
+    return "\u200B".join(text)
+
+
+def download_domain_list(bot: Sopel, path: str) -> bool:
+    """Download the current unsafe domain list.
+
+    :param path: Where to save the unsafe domain list
+    :returns: True if the list was updated
+    """
+    url = bot.settings.safety.domain_blocklist_url
+    if url is None or not url.startswith("http"):
+        url = "https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts"
+    LOGGER.info("Downloading unsafe domain list from %s", url)
+    old_etag = bot.db.get_plugin_value("safety", "unsafe_domain_list_etag")
+    if old_etag:
+        r = requests.head(url)
+        if r.headers["ETag"] == old_etag and os.path.isfile(path):
+            LOGGER.debug("Unsafe domain list unchanged, skipping")
+            return False
+
+    r = requests.get(url, stream=True)
+    try:
+        r.raise_for_status()
+        with open(path + ".new", "wb") as f:
+            for data in r.iter_content(None):
+                f.write(data)
+    except Exception:
+        # don't bother handling, we'll try again tomorrow
+        LOGGER.warning("Unsafe domain list download failed, using cache")
+        return False
+    # .new+move so we don't clobber it if the download fails in the middle
+    os.rename(path + ".new", path)
+    bot.db.set_plugin_value("safety", "unsafe_domain_list_etag", r.headers.get("etag"))
+    return True
+
+
+def update_local_cache(bot: Sopel, init: bool = False):
+    """Download the current malware domain list and load it into memory.
+
+    :param init: Load the file even if it's unchanged
+    """
+    path = os.path.join(bot.settings.homedir, "unsafedomains.txt")
+    updated = download_domain_list(bot, path)
+    if not os.path.isfile(path):
+        LOGGER.warning("Could not load unsafe domain list")
+        return
+
+    if not updated and not init:
         return
 
-    with open(loc, 'r') as f:
+    LOGGER.debug("Loading new unsafedomains list")
```

Comment:
Missing a space? Weird to have different logger calls use "unsafedomains list" and "unsafe domain list" interchangeably.